src/dhtproto/client/legacy/internal/helper/mirror/model/MirrorBase.d

```diff
@@ -0,0 +1,257 @@
+
+    /// Time (in milliseconds) to wait between successful GetAlls.
+    public Const!(uint) update_time_ms;
+
+    /// Time (in milliseconds) to wait between failed requests.
+    public Const!(uint) retry_time_ms;
+
+    /***************************************************************************
+
+        Constructor.
+
+        Params:
+            dht = DHT client used to access DHT
+            channel = name of DHT channel to mirror
+            update_time_s = seconds to wait between successful GetAlls
+            retry_time_s = seconds to wait between failed requests
+
+    ***************************************************************************/
+
+    public this ( Dht dht, cstring channel, uint update_time_s,
+        uint retry_time_s )
+    {
+        this.dht = dht;
+        this.channel_ = idup(channel);
+        this.update_time_ms = update_time_s * 1_000;
+        this.retry_time_ms = retry_time_s * 1_000;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            Name of DHT channel being mirrored
+
+    ***************************************************************************/
+
+    public istring channel ()
+    {
+        return this.channel_;
+    }
+
+    /// Struct, passed to start(), describing the requests to start.
+    public struct Setup
+    {
+        /// Enum describing the possible GetAll modes.
+        public enum GetAllMode
+        {
+            /// Do not assign GetAll requests.
+            None,
+            /// Assign one GetAll request but do not repeat.
+            OneShot,
+            /// Repeatedly assign GetAlls, according to the specified update
+            /// time.
+            Repeating,
+            /// Ditto, but assign the first immediately (no delay).
+            RepeatingNow
+        }
+
+        /// Enum describing the possible Listen modes.
+        public enum ListenMode
+        {
+            /// Do not assign Listen requests.
+            None,
+            /// Assign Listen requests (runs continuously).
+            Continuous
+        }
+
+        /// The desired GetAll mode. (Defaults to a repeating GetAll.)
+        GetAllMode get_all_mode = GetAllMode.Repeating;
+
+        /// The desired Listen mode. (Defaults to a continuous Listen.)
+        ListenMode listen_mode = ListenMode.Continuous;
+    }
+
+    /***************************************************************************
+
+        Assigns GetAll and/or Listen requests according to the specific setup.
+
+        Params:
+            setup = struct instance describing the requests to start (defaults
+                to a repeating GetAll request and a continuous Listen)
+
+    ***************************************************************************/
+
+    public void start ( Setup setup = Setup.init )
+    {
+        assert(!(setup.get_all_mode == setup.get_all_mode.None &&
+            setup.listen_mode == setup.listen_mode.None),
```

Comment:
Apply De-Morgan's law here and get: `setup.get_all_mode != setup.get_all_mode.None || setup_listen_mode != setup.listen_mode.None` feels more natural.

Reply:
Updated.